popup/search-results.js

```diff
@@ -0,0 +1,547 @@
+ * @returns {Object} Exposed methods representing the search results on userstyles.org
+ */
+function SearchUserstyles() {
+  let totalPages;
+  let currentPage = 1;
+  let exhausted = false;
+
+  return {getCurrentPage, getTotalPages, getCategory, isExhausted, search, fetchStyleJson, fetchStyle};
+
+  /**
+   * @returns {Number} The *Next* page to fetch for styles. Auto-incremented after each search.
+   */
+  function getCurrentPage() {
+    return currentPage;
+  }
+
+  /**
+   * @returns {Number} The total number of search result pages.
+   */
+  function getTotalPages() {
+    return totalPages;
+  }
+
+  /**
+   * @returns {Boolean} If there are no more results to fetch from userstyles.org
+   */
+  function isExhausted() {
+    return exhausted;
+  }
+
+  /**
+   * Resolves the Userstyles.org "category" for a given URL.
+   * @param {String} url The URL to a webpage.
+   * @returns {Promise<String>} The category for a URL, or the hostname if category is not found.
+   */
+  function getCategory(url) {
+    const hostname = new URL(url).hostname;
+    return new Promise(resolve => {
+      const request = new XMLHttpRequest();
+      const browseURL = 'https://userstyles.org/styles/browse/all/' + encodeURIComponent(url);
+      request.open('HEAD', browseURL, true);
+      request.onreadystatechange = () => {
+        if (request.readyState === XMLHttpRequest.DONE) {
+          const responseURL = new URL(request.responseURL);
+          const category = responseURL.searchParams.get('category');
+          if (category !== null) {
+            resolve(category);
+          } else {
+            resolve(hostname);
+          }
+        }
+      };
+      request.send(null);
+    });
+  }
+
+  /**
+   * Fetches the JSON style object from userstyles.org (containing code, sections, updateUrl, etc).
+   * This is fetched from the /styles/chrome/ID.json endpoint.
+   * @param {number} userstylesId The internal "ID" for a style on userstyles.org
+   * @returns {Promise<Object>} The response as a JSON object.
+   */
+  function fetchStyleJson(userstylesId) {
+    return new Promise((resolve, reject) => {
+      const jsonUrl = 'https://userstyles.org/styles/chrome/' + userstylesId + '.json';
+      download(jsonUrl)
+        .then(responseText => {
+          resolve(tryJSONparse(responseText));
+        })
+        .catch(reject);
+    });
+  }
+
+  /**
+   * Fetches style information from userstyles.org's /api/v1/styles/{ID} API.
+   * @param {number} userstylesId The internal "ID" for a style on userstyles.org
+   * @returns {Promise<Object>} An object containing info about the style, e.g. name, author, etc.
+   */
+  function fetchStyle(userstylesId) {
+    return new Promise((resolve, reject) => {
+      const TIMEOUT = 10000;
+      const headers = {
+        'Content-type': 'application/json',
+        'Accept': '*/*'
+      };
+
+      const styleUrl = new URL('https://userstyles.org/api/v1/styles/' + userstylesId);
```

Comment:
the base URL should be a named constant at the top of `searchUserstyles` closure